bin/backup

```diff
@@ -0,0 +1,174 @@
+#! /usr/bin/env bash
+
+set -euo pipefail
+
+#### Detect Toolkit Project Root ####
+# if realpath is not available, create a semi-equivalent function
+command -v realpath >/dev/null 2>&1 || realpath() {
+  [[ $1 = /* ]] && echo "$1" || echo "$PWD/${1#./}"
+}
+SCRIPT_PATH="$(realpath "${BASH_SOURCE[0]}")"
+SCRIPT_DIR="$(dirname "$SCRIPT_PATH")"
+TOOLKIT_ROOT="$(realpath "$SCRIPT_DIR/..")"
+if [[ ! -d "$TOOLKIT_ROOT/bin" ]] || [[ ! -d "$TOOLKIT_ROOT/config" ]]; then
+  echo "ERROR: could not find root of overleaf-toolkit project (inferred project root as '$TOOLKIT_ROOT')"
+  exit 1
+fi
+
+TMP_ROOT_DIR="$TOOLKIT_ROOT/tmp"
+
+IS_SERVER_PRO="$(grep -q 'SERVER_PRO=true' \
+  "$TOOLKIT_ROOT/config/overleaf.rc" && echo 'true' || echo 'false')"
+
+function usage() {
+  cat <<EOF
+Usage: bin/backup
+
+Makes a backup of the data in this installation, and writes it to a
+timestamped tar.gz file in the ./backup/ directory.
+
+This file can then be consumed by the bin/restore script.
+
+EOF
+}
+
+function wait-for-mongo () {
+  while ! "$TOOLKIT_ROOT/bin/docker-compose" exec -T mongo \
+    mongo --eval "db.version()" \
+  > /dev/null; do echo '[mongo is not ready]' && sleep 1; done
+  echo '[mongo is ready]'
+}
+
+function create-tmp-dir () {
+  local now
+  now="$(date '+%F-%H%M%S')"
+  local random_part
+  random_part="$(head -c 8 /dev/urandom | md5sum | cut -c 1-4)"
+  if ! [[ -d "$TMP_ROOT_DIR" ]]; then
+    mkdir "$TMP_ROOT_DIR"
+  fi
+  local tmp_dir="$TMP_ROOT_DIR/backup-$now-$random_part"
+  if [[ -d "$tmp_dir" ]]; then
+    echo "Error: temp directory '$tmp_dir' already exists" >&2
+    exit 1
+  fi
+  mkdir -p "$tmp_dir/backup"
+  echo "$tmp_dir"
+}
+
+function get-container-name () {
+  local name="$1"
+  "$TOOLKIT_ROOT/bin/docker-compose" ps | grep "$name" | cut -d ' ' -f 1 | head -n 1
+}
+
+function dump-mongo () {
+  local tmp_dir="$1"
+  local mongo_tmp_dir="$tmp_dir/backup/mongo"
+  mkdir "$mongo_tmp_dir"
+
+  "$TOOLKIT_ROOT/bin/docker-compose" up -d mongo
```

Comment:
Can we handle the case where there is an external Mongo DB here? I think there will be people running without the internal DBs.

Reply:
Yeah, would be useful to have a `--skip-mongo` option for both backup/restore.

Reply:
In addition, if we go with external redis, it could also have a `--skip-redis`

Reply:
I was thinking that we could look at the config in some way and back up the external mongo/redis using the same mechanism. Given that we've already checked the prerequisites for making a backup (the service is stopped so nothing will be in-flight) it makes sense to roll up the backup from the external DB at the same time. WDYT?